### PR TITLE
Implement a new, high-level SDK for `Consumer` and `Producer`

### LINKED ISSUE #39: Connection string
Implement connection string for `IggyClient` to automatically connect and authenticate the client connection.

### PR DESCRIPTION
**Pull request by spetz**
Additionally, refactor `IggyClient` and improve consumer group handling on the server.

This also closes #39

### CODE REVIEW

File: sdk/src/client_provider.rs [hunk 1]
```diff
@@ -59,8 +61,15 @@ impl ClientProviderConfig {
                     client_address: args.quic_client_address,
                     server_address: args.quic_server_address,
                     server_name: args.quic_server_name,
-                    reconnection_retries: args.quic_reconnection_retries,
-                    reconnection_interval: args.quic_reconnection_interval,
+                    reconnection: TcpClientReconnectionConfig {
```
Comment (hubcio) [inline sdk/src/client_provider.rs]: move this to before line 58 `let reconnection = ...` and inside this struct just use reconnection

File: sdk/src/tcp/config.rs [hunk 2]
```diff
@@ -1,35 +1,56 @@
...
     pub server_address: String,
-    /// The number of retries when connecting to the server.
-    pub reconnection_retries: u32,
-    /// The interval between retries when connecting to the server.
-    pub reconnection_interval: u64,
     /// Whether to use TLS when connecting to the server.
     pub tls_enabled: bool,
     /// The domain to use for TLS when connecting to the server.
     pub tls_domain: String,
+    // Whether to automatically sign in when connecting.
```
Comment (hubcio) [inline sdk/src/tcp/config.rs]: use triple `/`

Comment (spetz) [reply sdk/src/tcp/config.rs]: Fixed

File: examples/src/new-sdk/producer/main.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    );
+    let client_provider_config = Arc::new(ClientProviderConfig::from_args(args.to_sdk_args())?);
+    let client = client_provider::get_raw_connected_client(client_provider_config).await?;
+    let client = IggyClient::builder().with_client(client).build()?;
+    let mut producer = client
+        .producer(&args.stream_id, &args.topic_id)?
+        .batch_size(args.messages_per_batch)
+        .send_interval(IggyDuration::from_str(&args.interval)?)
+        .partitioning(Partitioning::balanced())
+        .create_topic_if_not_exists(3, None)
```
Comment (numinnex) [inline examples/src/new-sdk/producer/main.rs:31]: Maybe we could rename this to `bootstrap_resources` and create everything that is required, stream including ? and make the partition count have a default.

File: sdk/src/clients/consumer.rs [hunk 4]
```diff
@@ -0,0 +1,743 @@
...
+use std::time::Duration;
+use tokio::time::sleep;
+use tracing::{error, info, trace, warn};
+
+const ORDERING: std::sync::atomic::Ordering = std::sync::atomic::Ordering::SeqCst;
+type PollMessagesFuture = Pin<Box<dyn Future<Output = Result<PolledMessages, IggyError>>>>;
+
+/// The auto-commit configuration for storing the offset on the server.
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum AutoCommit {
```
Comment (numinnex) [inline sdk/src/clients/consumer.rs:30]: maybe every N'th message mode would be useful ?

File: examples/src/new-sdk/producer/main.rs (hunk 3 above)
Comment (spetz) [reply examples/src/new-sdk/producer/main.rs:31]: These are not required, as they already have the default values in place :)

File: sdk/src/clients/consumer.rs (hunk 4 above)
Comment (spetz) [reply sdk/src/clients/consumer.rs:30]: Done

**Review by numinnex — DISMISSED** [review]

**Review by numinnex — DISMISSED** [review]

**Review by numinnex — APPROVED** [review]

**Pull request merged**